### PR TITLE
Snapshot to chainspec command

### PR DESCRIPTION
**Pull request by ControlCplusControlV**

### CODE REVIEW

**Comment by ControlCplusControlV** [comment]
Moved to PR so I can see everything to clean up easier, mostly working but still sorting out a cleaner test and potential state_config issue

**Comment by rakita** [comment]
I did a little bit of refactoring of the main config: https://github.com/FuelLabs/fuel-core/pull/440 just a heads up as there will be conflicts.

**Comment by ControlCplusControlV** [comment]
Ok, going to hold off requesting review until those are merged then

**Comment by ControlCplusControlV** [comment]
Converting to draft while I removed feature flags as requested

**Comment by ControlCplusControlV** [comment]
Closes #75

File: fuel-core/src/config/chain_config.rs
Comment (leviathanbeak) [inline fuel-core/src/config/chain_config.rs:126]: are these things always assumed to work or should we be more careful with `unwraps` here ?

File: fuel-core/src/database/coin.rs
Comment (leviathanbeak) [inline fuel-core/src/database/coin.rs:131]: few `unwraps` here as well - would it be better that we rather return an `error` instead of panicking here?

File: fuel-core/src/database/contracts.rs
Comment (leviathanbeak) [inline fuel-core/src/database/contracts.rs:80]: `Working?` comments here - what are they for :D

**Review by leviathanbeak — DISMISSED** [review]
I think there's a lot of `unwraps` everywhere, in methods that actually return `Result<_, anyhow::Error>` so it would be better to cast those errors into `anyhow::Error` and return them rather than potentially `unwrap()` panicking in the middle of the execution.

File: fuel-core/src/database/coin.rs
Comment (ControlCplusControlV) [reply fuel-core/src/database/coin.rs:131]: I initialing had trouble unwrapping inside of that closure because I wasn't returning an error, but I agree that it is dangerous and will adjust this to properly propagate errors to the top

File: fuel-core/src/config/chain_config.rs
Comment (ControlCplusControlV) [reply fuel-core/src/config/chain_config.rs:126]: Moving this to `unwrap_or_default`, these should always work and properly return a `None` value if nothing exists, but moving to `unwrap_or_default` so an error in one of them doesn't ruin the generation of the entire config

**Comment by ControlCplusControlV** [comment]
> I think there's a lot of `unwraps` everywhere, in methods that actually return `Result<_, anyhow::Error>` so it would be better to cast those errors into `anyhow::Error` and return them rather than potentially `unwrap()` panicking in the middle of the execution.

I agree, I moved both coin and contract config generation to now `unwrap` much smarter and propagate the error upward if one does occur. I also dropped in better error handling in `state_config` generation by falling back to a default value in an error occurs so the rest of the config can still be exported in the event there is in error in one part of it.

File: fuel-core/src/args.rs
Comment (Voxelot) [inline fuel-core/src/args.rs:34]: [nit] any reason for the `_` prefix here?

File: fuel-core/src/args.rs
Comment (Voxelot) [inline fuel-core/src/args.rs:193]:
```suggestion
        let db = Database::open(&path).context(|| format!("failed to open database at path {}", path))?;
```

File: fuel-core/src/args.rs
Comment (Voxelot) [inline fuel-core/src/args.rs:207]: 1. No need to unwrap here when we have the convenience of anyhow available
2. Consider using streaming serialization instead of copying the whole string into memory

```suggestion
        let mut stdout = io::stdout().lock();
        let serialized = serde_json::to_writer(stdout, &chain_conf).context("failed to dump snapshot to JSON")?;
```

File: fuel-core/src/database/coin.rs
Comment (Voxelot) [inline fuel-core/src/database/coin.rs:131]: is unwrap needed here?

File: fuel-core/src/database/coin.rs
Comment (Voxelot) [inline fuel-core/src/database/coin.rs:140]: more unwrap

File: fuel-core/src/database/contracts.rs
Comment (Voxelot) [inline fuel-core/src/database/contracts.rs:90]: `get` only fetches a single key, we need to use `iter_all` to collect _all_ the state of the contract. Also these columns use a `MultiKey`, where contract id is just the prefix to the actual state key.

File: fuel-tests/tests/snapshot.rs
Comment (Voxelot) [inline fuel-tests/tests/snapshot.rs:33]: test should cover state and balances as well, as it would reveal current problems in the implementation

File: fuel-core/src/args.rs
Comment (ControlCplusControlV) [reply fuel-core/src/args.rs:182]: config is unused if rocksdb isn't enabled because a snapshot couldn't be taken, and so it's never used. I got around this with path by printing it in the error, would you like me to do the same with config?

File: fuel-core/src/args.rs
Comment (ControlCplusControlV) [reply fuel-core/src/args.rs:207]: I added this in but not commit suggestion because it actually further can have some changes such to make this look like
```rust
        let stdout = io::stdout().lock();
        serde_json::to_writer(stdout, &chain_conf).context("failed to dump snapshot to JSON")?;
        ```

File: fuel-core/src/args.rs
Comment (Voxelot) [reply fuel-core/src/args.rs:182]: Ah makes sense, I think it's fine as is then.

File: fuel-tests/Cargo.toml
Comment (Voxelot) [inline fuel-tests/Cargo.toml:38]: This doesn't seem necessary now that we removed full integ testing of the CLI

File: fuel-tests/Cargo.toml
Comment (Voxelot) [inline fuel-tests/Cargo.toml:29]: Also not needed anymore

File: fuel-tests/Cargo.toml
Comment (ControlCplusControlV) [reply fuel-tests/Cargo.toml:38]: will remove

File: fuel-tests/Cargo.toml
Comment (ControlCplusControlV) [reply fuel-tests/Cargo.toml:29]: will remove

File: fuel-core/src/args.rs
Comment (rakita) [reply fuel-core/src/args.rs:29]: https://github.com/FuelLabs/fuel-core/issues/457

File: fuel-core/src/args.rs
Comment (rakita) [inline fuel-core/src/args.rs:80]: [nit] This probably deserved it's own mod, as in `args/snapshot.rs`, can be put under effort of refactoring cli subcommand: https://github.com/FuelLabs/fuel-core/issues/457

File: fuel-core/src/args.rs
Comment (rakita) [reply fuel-core/src/args.rs:179]: Usually, you specify two functions one with an error and one without it and use cfg to enable/disable them.

File: fuel-core/src/args.rs
Comment (ControlCplusControlV) [reply fuel-core/src/args.rs:29]: I was wondering if it was worth re-vamping the cli structure with this PR but wanted to keep changes minimal. I definitely agree though because this implementation is a bit clunky

File: fuel-core/src/database/coin.rs
Comment (ControlCplusControlV) [reply fuel-core/src/database/coin.rs:140]: This is an option Unwrap not a result unwrap, so should be safe right?

File: fuel-core/src/args/run.rs
Comment (Voxelot) [inline fuel-core/src/args/run.rs:97]: As this logging setup could be used for both snapshot and running, it should be moved to the central `run_cli` rather than only happening when a node is run.

File: fuel-core/src/database/coin.rs
Comment (Voxelot) [reply fuel-core/src/database/coin.rs:140]: Why do we need to fetch the coin by ID if we've already unwrapped it from `raw_coin` above?

File: fuel-core/src/args/run.rs
Comment (ControlCplusControlV) [reply fuel-core/src/args/run.rs]: Moved it into the outer `args.rs` so it should be accessible for subcommands, and starts regardless of the subcommand being run

File: fuel-tests/tests/snapshot.rs
Comment (Voxelot) [inline fuel-tests/tests/snapshot.rs:90]: We should have assertions that the state key, state values, and balances are correct, and not just set to any random value. Perhaps we should split this test up across the various areas of concern, such as a test for each of these:
- contract deployments (eg. bytecode, salt)
- contract balances
- contract state
- coins
- block height

These tests should verify the exact expected values.

**Comment by ControlCplusControlV** [comment]
I removed the expensive and ugly double collection, and then remove the extra get within the balances state collection (fixing what @Voxelot had asked in using all of the values provided in `iter_all`)

File: fuel-tests/tests/snapshot.rs
Comment (ControlCplusControlV) [reply fuel-tests/tests/snapshot.rs:90]: Ok, will expand this test. Sorry didn't see this until I had just re-requested review in the 2nd commit

File: fuel-core/src/config/chain_config.rs
Comment (Voxelot) [inline fuel-core/src/config/chain_config.rs:128]: This method should propagate the errors from the database methods without unwrap. Otherwise, the snapshot could silently fail while still providing partially valid results without the user knowing.

```suggestion
    pub fn generate_state_config(db: Database) -> anyhow::Result<Self> {
        StateConfig {
            coins: db.get_coin_config()?,
            contracts: db.get_contract_config()?,
            height: db.get_block_height()?,
```

Comment (ControlCplusControlV) [reply fuel-core/src/config/chain_config.rs:128]: I initially did this so corruption in one column would allow the rest to be saved, but I see how this silent failure could be more harmful

Comment (Voxelot) [reply fuel-core/src/config/chain_config.rs]: This could be enabled in the future using a `continue-on-error` flag, however, any errors should be reported via `error!` or `warn!` before `unwrap_or_default` so the user is aware that their snapshot may come with some issues.

**Review by Voxelot — APPROVED** [review]

**Pull request merged**
